/*
 * ProfaneDB - A Protocol Buffers database.
 * Copyright (C) 2017  "Giorgio Azzinnaro" <giorgio.azzinnaro@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

syntax = "proto3";

package profanedb.protobuf;

option go_package = "gitlab.com/profanedb/protobuf/db";
option csharp_namespace = "ProfaneDB.Protobuf";
option java_package = "com.profanedb.protobuf";
option objc_class_prefix = "PDB";

import "google/protobuf/any.proto";

// A Key uniquely identifies a message stored in the database
message Key {
	string message_type = 1; // A identifier for the message
	string field = 2; // The field within the message
	bytes value = 3; // The value serialized in such way that it doesn't create collisions
}

// A StorableMessage has a unique Key and a serialized representation of the object
message StorableMessage {
  Key key = 1;
  google.protobuf.Any payload = 2; // This gives much more flexibility, as it can be read again
}

// A Message might depend on other messages to be stored first
message MessageTreeNode {
  StorableMessage message = 1;
  repeated MessageTreeNode children = 2;
}
